### ISSUE TITLE
A new "allowBranches" regular expression to work with "NewestFromBranch" with git warehouse subscriptions

### ISSUE THREAD

**Issue by kennethleider** [akuity/kargo#2907]
# Checklist

* [X] I've searched the issue queue to verify this is not a duplicate feature request.
* [X] I've pasted the output of `kargo version`, if applicable.
* * v1.0.3
* [X] I've pasted logs, if applicable.

# Proposed Feature

I would like the "NewestFromBranch" commitSelectionStrategy to monitor more than just one branch

# Motivation

We develop in feature branches, and I would like to be able to subscribe to freight in any active feature branch.  Since we follow a naming convention for the branch, it would be fairly simple to detect changes in just these branches

# Suggested Implementation

You could either modify "allowTags" to also evaluate against branch names when the commitSelectionStrategy is "NewestFromBranch", or add a new "allowBranches" that works the same way.

**Comment by krancour** [comment]
Thank you for the suggestion @kennethleider.

Can you elaborate on how you see this working?

For image repos, the various selection strategies select _A_ (singular) image version. For git repos, they select _A_ (singular) commit.

Selecting _one_ newest commit from n > 1 branches seems like it wouldn't be that useful. "Grab the latest commit from among these n branches?" It's a little bit of a crap-shoot as to what you're going to get...

So, reading between the lines, I can surmise that what you actually want is for n branches, you want to select _up to_ n newest commits (_not just one_; i.e. newest previously undetected commit from _each_). Have I got that right?

This would amount to a much bigger change than just a few tweaks to a selection strategy.

As I'm having a difficult time imagining how a Freightline with manifests from many branches wouldn't promote some confusion that harms the overall UX, some more detail on how you would foresee yourself actually using this feature end-to-end would help a great deal.

**Comment by kennethleider** [comment]
>  one newest commit from n > 1 branches 

Nope this is exactly what I want.  I think you are imaging that I have multiple active branches.  It is more like I have a branch for each release of the product.  Issues in the release are merged into the branch, at which point I want to be able to promote them.  Once the release is done, it is merged back into main, and a new release branch is created the next time work needs to be done on the application.

It would not be a big deal to update the warehouse to have the name of the branch when ever we start a new release, but I would prefer that it just pick up any branch that matches a branch name pattern.

**Comment by Brightside56** [comment]
I was planning to use commits from more than one branch in order to allow engineers to test changes from feature branches on certain stages

**Comment by krancour** [comment]
@kennethleider your explanation was very enlightening. Thank you!

I worry a little that selecting the newest commit from among multiple branches could lead to some surprises if something is pushed to a branch that matches the pattern but _isn't_ the one you currently regard as the "active" branch.

> It would not be a big deal to update the warehouse to have the name of the branch when ever we start a new release

For now, this would be my suggestion until/unless we are able to figure out the mechanics here that would make something like this work nicely while still safeguarding against the surprises/confusion I am concerned about.
